File: src_fortran/sub/datum.f
c
c
c
c
c
c
      subroutine DATUM(ITF,IYR,IMO,IDY,IHR,IMN)
C UMRECHNEN DES DATUMS IN MINUTEN (CF. JULIAM) IN YR-MO-DY-HR-MI
C   (MIT IMN<2**31, JAHR < 4000
c
      implicit none
      integer iyr,imo,idy,ihr,imn
c      
      integer id,l,iyr4,iyrh,iyrt,ld,i
      integer KMO(12)
      INTEGER ITF,K,KH
      DATA KMO/31,28,31,30,31,30,31,31,30,31,30,31/
      K= ITF/60
      IMN= ITF-K*60
      KH= K/24
      IHR= K-KH*24
      IYR= KH/365
5     ID= KH-IYR*365
      L= 0
      IYR4= IYR/4
      IYRH= IYR/100
      IYRT= IYR/1000
      LD= IYR4-IYRH+IYRT
      IF(IYR4*4.EQ.IYR.AND.(IYRH*100.NE.IYR.OR.IYRT*1000.EQ.IYR)) L= 1
      ID= ID-LD+L
      IF(ID.GT.0) GOTO 10
      if(id.eq.0.and.ihr.eq.0.and.imn.eq.0) then
          idy= 0
          imo= 0
          return
      endif
      IYR= IYR-1
      GOTO 5
10    KMO(2)= 28+L
      DO 20 I=1,12
      ID= ID- KMO(I)
      IF(ID.LE.0) GOTO 30
20    CONTINUE
      I=12
30    IDY= ID+KMO(I)
      IMO= I
      RETURN
      end ! of subr. datum
c
      integer function TRIMLEN(t)   ! Urs Kradolfer, June 1986
c
c     Call:    nc=TRIMLEN(char)
c
c          --> nc says, how many characters the input-string has
c              (ignoring trailing blanks!).
c
      implicit none
      character t*(*)
      do 1 trimlen=LEN(t),1,-1
    1    if(t(trimlen:trimlen).ne.' ')RETURN
      trimlen=1
      end ! of integer function trimlen

File: src/sub/datum.c
#include <stddef.h>

int trimlen(const char *t) {
    int len;

    if (t == NULL) {
        return 1;
    }

    for (len = 0; t[len]; ++len) {
    }

    while (len > 0 && t[len - 1] == ' ') {
        --len;
    }

    return len > 0 ? len : 1;
}

void datum(int itf, int *iyr, int *imo, int *idy, int *ihr, int *imn) {
    int kmo[12] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
    int k, kh, id, l, iyr4, iyrh, iyrt, ld, i;

    if (iyr == NULL || imo == NULL || idy == NULL || ihr == NULL ||
        imn == NULL) {
        return;
    }

    k = itf / 60;
    *imn = itf - k * 60;
    kh = k / 24;
    *ihr = k - kh * 24;
    *iyr = kh / 365;

    while (1) {
        id = kh - *iyr * 365;
        l = 0;
        iyr4 = *iyr / 4;
        iyrh = *iyr / 100;
        iyrt = *iyr / 1000;
        ld = iyr4 - iyrh + iyrt;

        if (iyr4 * 4 == *iyr && (iyrh * 100 != *iyr || iyrt * 1000 == *iyr)) {
            l = 1;
        }

        id = id - ld + l;

        if (id > 0) {
            break;
        }

        if (id == 0 && *ihr == 0 && *imn == 0) {
            *idy = 0;
            *imo = 0;
            return;
        }

        *iyr = *iyr - 1;
    }

    kmo[1] = 28 + l;

    for (i = 0; i < 12; ++i) {
        id = id - kmo[i];
        if (id <= 0) {
            break;
        }
    }

    if (i == 12) {
        i = 11;
    }

    *idy = id + kmo[i];
    *imo = i + 1;
}
